realm/realm-library/src/main/java/io/realm/internal/android/JsonUtils.java

```diff
@@ -37,13 +40,25 @@
      */
     public static Date stringToDate(String date) {
         if (date == null || date.length() == 0) return null;
+
+        // Check for JSON date
         Matcher matcher = jsonDate.matcher(date);
         if (matcher.find()) {
             String dateMatch = matcher.group(1);
             return new Date(Long.parseLong(dateMatch));
-        } else {
+        }
+
+        // Check for millisecond based date
+        if(numericOnly.matcher(date).matches()) {
             return new Date(Long.parseLong(date));
         }
+
+        // Try for ISO8601 date
+        try {
+            return ISO8601Utils.parse(date, new ParsePosition(0));
```

Comment:
`new ParsePosition(0)` should be a static variable. No need to allocate it each time.